### PR TITLE
KAFKA-15416: Fix flaky TopicAdminTest::retryEndOffsetsShouldRetryWhenTopicNotFound test case

### PR DESCRIPTION
**Pull request by C0urante**
[Jira](https://issues.apache.org/jira/browse/KAFKA-15416)

There's some fairly subtle behavior involved here, so apologies in advance for the wall of text.

# Problem

In the `retryEndOffsetsShouldRetryWhenTopicNotFound` test case, we currently use a `MockTime` object that has an auto-tick of 10ms with our [AdminClientUnitTestEnv](https://github.com/apache/kafka/blob/703e1d9faafbf07795261b3233ab985583f17fcb/clients/src/test/java/org/apache/kafka/clients/admin/AdminClientUnitTestEnv.java), which ends up also being utilized by the [admin client](https://github.com/apache/kafka/blob/703e1d9faafbf07795261b3233ab985583f17fcb/clients/src/test/java/org/apache/kafka/clients/admin/AdminClientUnitTestEnv.java#L107-L109) that it provides.

By default, admin clients perform periodic metadata refresh every five minutes (see the [metadata.max.age.ms property](https://github.com/apache/kafka/blob/703e1d9faafbf07795261b3233ab985583f17fcb/clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java#L137)). With a `MockTime` instance that has an auto-tick of 10ms, periodic metadata refresh will be triggered every 30,000 times that `MockTime::milliseconds` (or `MockTime::nanoseconds`, but that's not relevant to these tests) is called.

Because of how the testing admin is set up by the `AdminClientUnitTestEnv` class, [checks for periodic metadata refresh](https://github.com/apache/kafka/blob/703e1d9faafbf07795261b3233ab985583f17fcb/clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java#L1359-L1368) take place in a tight loop with no time spent sleeping during each iteration. As a result, it's not unreasonable for 30,000 iterations to take place fairly quickly.

When a periodic metadata refresh is triggered, it can sometimes "swallow" the responses that we [prepare for the admin client](https://github.com/apache/kafka/blob/703e1d9faafbf07795261b3233ab985583f17fcb/connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java#L563-L565), causing these responses to be delivered to the periodic metadata refresh request instead of requests that the admin client issues in service of application logic (e.g., an invocation of `Admin::listOffsets`).

As a brief refresher on admin client behavior, when servicing a call to `Admin::listOffsets`, on the happy path, two broker requests are made: the first is a metadata request for the specified topic partitions, and the second is a request for the actual offset info of the requested topic partitions.

With the current `TopicAdminTest::retryEndOffsetsShouldRetryWhenTopicNotFound` case, this can lead to a variety of timing-dependent failures:

- If every prepared response is swallowed, a spurious timeout is caused by repeated attempts to contact a node that never responds because all prepared responses have already been consumed, with every attempt causing the mocked time to move forward slowly but surely past the operation timeout
- If two prepared responses are swallowed before `Admin::listOffsets` is invoked, a `ClassCastException` is thrown because the metadata request issued in service of the call to `Admin::listOffsets` receives the response prepared for the list offsets request, and the admin (somewhat reasonably) [expects the response for that metadata request to be a `MetadataResponse` instance](https://github.com/apache/kafka/blob/703e1d9faafbf07795261b3233ab985583f17fcb/clients/src/main/java/org/apache/kafka/clients/admin/internals/PartitionLeaderStrategy.java#L144)
- If a single response is swallowed, the test passes, but the test is rendered ineffective since the retry logic it intends to verify is never actually covered

# Solutions

There are a few options for a fix:

1. Disable periodic metadata refresh in the admin client
1. Disable auto-tick in our `MockTime` instance
1. Switch to a fully-deterministic mocking method for the admin client (e.g., use Mockito to construct a mock `Admin` instance)

In this PR I've opted for the second option, as this appears to be (implicitly or explicitly) the strategy used in all other tests that use the `AdminClientUnitTestEnv` class, and I'd prefer not to introduce inconsistency in the strategies we use in the `TopicAdminTest` suite to mock out admin client behavior.

# Other changes

## Test retryEndOffsetsShouldWrapNonRetriableExceptionsWithConnectException

This PR also tweaks the `retryEndOffsetsShouldWrapNonRetriableExceptionsWithConnectException` test case to prevent false negatives. Currently, because of a resurfacing of [KAFKA-12879](https://issues.apache.org/jira/browse/KAFKA-12879), admin clients will retry attempts to list offsets on unknown topics. The test expects the opposite behavior, but because the next automatic retry is met with a spurious timeout, it fails anyways.

The test is modified here to 1) prepare a different failure (authorization error) that is less likely to be be treated as retriable by future changes, 2) prepare a valid response for the admin client in the event that it ever retries the operation (which should, conversely, cause the test to fail), and 3) make finer-grained assertions on the kind of exception thrown by `TopicAdmin::retryEndOffsets` to ensure that it fails for the reason that we expect it to.

## TopicAdmin::retryEndOffsets and TopicAdmin::endOffsets

A few tweaks are made to these methods to eliminate unnecessary chained elements in the stack trace of exceptions they throw. This not only facilitates the stronger testing assertions added to `TopicAdminTest::retryEndOffsetsShouldWrapNonRetriableExceptionsWithConnectException`, it also improves the readability of the stack traces shown to users.

This same improvement may be made to other parts of the `TopicAdmin` class. I've left that for follow-up work to try to keep this PR focused for now, but can add it if we agree it's warranted.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by C0urante** [comment]
@philipnee I did my best to research the context here diligently, but I may be missing something (this is a bit outside my wheelhouse). If you have a moment, would you be willing to take a look?

**Comment by philipnee** [comment]
hey @C0urante - I'm looking into this.  I can't seem to reproduce the error you reported on the jira, and i've tried different JDK versions.  I'll follow up here if i'm able to make some progress.

**Review by gharris1727 — APPROVED** [review]
LGTM, thanks @C0urante for the fix!

File: connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java
Comment (gharris1727) [inline connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java]: This wasn't introduced in this PR but it did draw my attention to it:

This Object argument for TopicAdmin feels very strange to me, and I see that it is only used in log messages, so it is left as `null` in most tests. Perhaps we can follow the `null` convention in this test, and in a follow-up try to eliminate the Object argument.

Comment (C0urante) [reply connect/runtime/src/test/java/org/apache/kafka/connect/util/TopicAdminTest.java]: Good point! If memory serves, I originally went with the `Object` parameter in the `TopicAdmin` constructor for ergonomic reasons: the configuration accepted by `Admin::create` is a `Map<String, Object>`, so it would be a simple matter of passing something like `adminConfig.get("bootstrap.servers")` as the first argument instead of having to do pesky null-sensitive `toString` operations.

But I agree that this isn't especially elegant or, more importantly, safe: there are even cases (thankfully only in tests) where we passed the entire config object as the `bootstrapServers` argument instead of just a string.

I've pushed a commit that refactors the `TopicAdmin` constructors and all applicable call sites to be more type-safe.

If this is easy enough to review, LMK and we can include it in this PR. Otherwise, I'm happy to revert the commit, merge this PR, and then file a follow-up to clean up the `TopicAdmin` constructors.

**Comment by C0urante** [comment]
@philipnee If it helps, this test only fails for me about one out of every three runs, and only when running the entire suite of unit tests for `:connect:runtime`--I've never seen it fail when run individually.

I've also discussed offline with @gharris1727 and he's never seen this test fail either. Guess my laptop is just cursed? 🤔

If it's too difficult to reproduce, it'd still be helpful to get your assessment on the analysis in the description and whether it passes the smell test and/or goes against intentional decisions made when these tests were originally written. One thing that may be useful to know if you can recall is why the `MockTime` instance for these tests had a non-zero auto-tick?

**Comment by C0urante** [comment]
@philipnee Let me know if you're still looking into this. Happy to wait on your analysis if you are, and if not, we can merge as-is.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java
Comment (philipnee) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:298]: sorry for being a type police 😆 : But this logic annoys me

```
public TopicAdmin(Object bootstrapServers, Admin adminClient) {
        this(bootstrapServers, adminClient, true);
    }

    // visible for testing
    TopicAdmin(Object bootstrapServers, Admin adminClient, boolean logCreation) {
        this.admin = adminClient;
        this.bootstrapServers = bootstrapServers != null ? bootstrapServers.toString() : "<unknown>";
        this.logCreation = logCreation;
    }
```

Can we just do map.getOrDefault() and perform an inline type cast? In general passing nulls around confuses me...

Comment (philipnee) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:298]: also understood this is not your code ... So I can submit a minor patch for this if you don't want to modify your PR.

**Review by philipnee — APPROVED** [review]
Hey @C0urante - The PR looks good and thanks for refactoring the tests.  To your question: I'm also not sure why I originally used a non-zero auto-tick, I think it was probably because of lack of familiarity with the code base at the time of writing the PR.

However, I haven't yet been able to reproduce this error locally, so i'm still unclear about the cause of the flakiness. in fact, there are quite a few tests like this in the client code base but no one seems to report them.

I left 1 comment.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:298]: I'm not a huge fan of unchecked type casts when I can avoid them, and the default of `"<unknown>"` that we use in the root constructor should alleviate any concerns about NPEs. I guess we could maybe do something like `adminConfig.getOrDefault(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, "<unknown>").toString()` (though this would be susceptible to NPEs if `adminConfig` contained a null value for that key), but I agree that it's not a blocker.

Happy to review if you want to submit that follow-up patch, though!

**Comment by C0urante** [comment]
@philipnee I'm curious about the other tests as well. From what I could tell, the distinguishing factor is the use of a `MockTime` instance that has a non-zero auto-tick, in conjunction with the `AdminClientUnitTestEnv`. I couldn't find any other instances of that in our tests; did any escape me?

**Comment by philipnee** [comment]
@C0urante - Sorry what i meant was I've seen other locally flaky tests that almost never show up during the jenkins build.   Thanks! I'll submit a patch after you merge it.

**Comment by C0urante** [comment]
Ah, that makes more sense. Yeah, I've seen some of those too, would be nice to patch them. Thanks for your help on this one @philipnee!

**Comment by philipnee** [comment]
@C0urante - Thank you!

**Pull request merged**
